rusoto/credential/src/environment.rs

```diff
@@ -36,35 +46,173 @@ impl ProvideAwsCredentials for EnvironmentProvider {
 
 /// Grabs the Credentials from the environment. These credentials are good for 10 minutes.
 fn credentials_from_environment() -> Result<AwsCredentials, CredentialsError> {
-    let env_key = match env_var("AWS_ACCESS_KEY_ID") {
-        Ok(val) => val,
-        Err(_) => return Err(CredentialsError::new("No AWS_ACCESS_KEY_ID in environment")),
+    let env_key = match non_empty_env_var(AWS_ACCESS_KEY_ID) {
+        Some(val) => val,
+        None => return Err(CredentialsError::new(E_NO_ACCESS_KEY_ID)),
     };
-    let env_secret = match env_var("AWS_SECRET_ACCESS_KEY") {
-        Ok(val) => val,
-        Err(_) => {
-            return Err(CredentialsError::new(
-                "No AWS_SECRET_ACCESS_KEY in environment",
-            ))
+    let env_secret = match non_empty_env_var(AWS_SECRET_ACCESS_KEY) {
+        Some(val) => val,
+        None => {
+            return Err(CredentialsError::new(E_NO_SECRET_ACCESS_KEY))
         }
     };
-
-    if env_key.is_empty() || env_secret.is_empty() {
-        return Err(CredentialsError::new(
-            "Couldn't find either AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY or both in environment.",
-        ));
-    }
-
     // Present when using temporary credentials, e.g. on Lambda with IAM roles
-    let token = match env_var("AWS_SESSION_TOKEN") {
-        Ok(val) => if val.is_empty() { None } else { Some(val) },
-        Err(_) => None,
+    let token = non_empty_env_var(AWS_SESSION_TOKEN);
+    // Mimik botocore's behavior, see https://github.com/boto/botocore/pull/1187.
+    let expires_at = match non_empty_env_var(AWS_CREDENTIAL_EXPIRATION) {
+        Some(val) => match DateTime::<Utc>::from_str(&val) {
+            Ok(e) => Some(e),
+            Err(e) => return Err(CredentialsError::new(format!("{} '{}': {}", E_INVALID_EXPIRATION, val, e)))
+        },
+        _ => None,
```

Comment:
I'm behind on understanding our new async/futures based credential sourcing. Is this statement correct? We don't have to specify an `expires_at` because `impl ProvideAwsCredentials for InstanceMetadataProvider`'s `fn credentials(&self) -> Self::Future` method delegates polling to other code we have that checks for expired credentials.

If that is correct, using an IAM instance profile will still correctly handle the env vars rotating to new credential keys.

Reply:
Hmm, I might be wrong, but combined with the changes to `credential/src/lib` in this PR, I believe that this would cause the credentials obtained via `EnvironmentProvider` to never be refreshed.

Setting up an IAM instance profile should result in the credentials to be provided by `InstanceMetadataProvider`.

Or am I misunderstanding your question?

Reply:
I think we're talking about the same thing, yeah. 😄 Rephrased: it *looks* like instance profile credentials won't refresh but I'm unable to confirm from reading the code.

Reply:
This change only affects the Environment provider and extends it to the same behavior as botocore. Note, ```AWS_CREDENTIAL_EXPIRATION``` is optional. This should not affect instance profile provider at all. Credentials retrieved from instance metadata or ECS will still be rotated.